classes/DirtEventTypes.sc

```diff
@@ -38,89 +38,93 @@ DirtEventTypes {
 	midiEvent = (
 			play: #{
 
-				var freq, lag, sustain, func;
-				var args, midiout, hasGate, midicmd, latency;
-				var sendNRPN;
+				var freq, lag, sustain;
+				var args, midiout, hasGate, midicmd, latency, chan;
+				var sendNRPN, schedmidi, schedmidicmd;
+				var hasNote = ~n != \none;
 
-				midicmd = ~midicmd;
 				midiout = ~midiout.value;
 
-				if(midicmd.isNil) {
-					if(~ccn.notNil) { midicmd = \control; ~ctlNum = ~ccn };
-					if(~ccv.notNil) { midicmd = \control; ~control = ~ccv };
-					if(~nrpn.notNil) {
-						var chan, nrpnMSB, nrpnLSB, valMSB, valLSB;
-						~val = ~val ? 0;
-						chan = ~midichan ? 0;
-						midicmd = \control;
-						nrpnLSB = ~nrpn % 128;
-						nrpnMSB = (~nrpn - nrpnLSB) / 128;
-						valLSB  = ~val % 128;
-						valMSB  = ((~val - valLSB) / 128) % 128;
-						sendNRPN = {
-							midiout.control(chan, 99, nrpnMSB);
-							midiout.control(chan, 98, nrpnLSB);
-							midiout.control(chan, 6,  valMSB);
-							midiout.control(chan, 38, valLSB)
-						};
-					};
-					if(~progNum.notNil) { midicmd = \program };
-					if(~polyTouch.notNil) { midicmd = \polyTouch };
-					if(~midibend.notNil) { midicmd = \bend; ~val = ~midibend; };
-					if(~miditouch.notNil) { midicmd = \touch; ~val = ~miditouch; };
-					if(midicmd.isNil) { midicmd = \noteOn }; // if still nil
+				if(midiout.isNil) {
+					"midi device is nil.".postln;
+					^true
```

Comment:
better not exit like this. Also `true` doesn't make sense as a return value, because in all the other cases, the event returns itself. So when you call:
```
x = event.play
```
you sometimes get an `Event` for x, sometimes `true`. Why not throw an error?

Reply:
hm, on a second thought, you may not want to flood the post window with heaps of error messages. I'll think of an error that you can throw and that won't do that.

Reply:
Ok, done. Now you can add an error to the midievent that you can then throw.

```supercollider
midiEvent = (
			midiOutNotFoundError: DirtPartTimeError(2, "midi device is nil."),
			play: #{

			<…>
			if(midiout.isNil) {
					~midiOutNotFoundError.throw;
			}


```

Reply:
Thanks, I've pushed the change, but it's not working for me, `DirtPartTimeError` doesn't seem to exist:
```
Execution warning: Class 'DirtPartTimeError' not found
ERROR: Message 'new' not understood.
```

Reply:
sorry, I forgot to merge. Now it should work.

Reply:
Great, that's better. It generates a large call stack error dump, but otherwise works fine.

Reply:
is it bad that it is a large dump? That's usually the way it is done for "proper" errors, i.e. those you can't ignore

Reply:
I find these stack dumps difficult to interpret and in this case the information isn't useful. I think it could be a problem to fill the postbuffer in a performance, if I mistyped the name of a midi device.